modules/abstract-utxo/src/abstractUtxoCoin.ts

```diff
@@ -1086,40 +1116,73 @@ export abstract class AbstractUtxoCoin extends BaseCoin {
       isLastSignature = params.isLastSignature;
     }
 
-    if (_.isUndefined(userPrv) || !_.isString(userPrv)) {
-      if (!_.isUndefined(userPrv)) {
-        throw new Error(`prv must be a string, got type ${typeof userPrv}`);
+    const getSignerKeychain = (): utxolib.BIP32Interface => {
+      const userPrv = params.prv;
+      if (_.isUndefined(userPrv) || !_.isString(userPrv)) {
+        if (!_.isUndefined(userPrv)) {
+          throw new Error(`prv must be a string, got type ${typeof userPrv}`);
+        }
+        throw new Error('missing prv parameter to sign transaction');
       }
-      throw new Error('missing prv parameter to sign transaction');
-    }
+      const signerKeychain = bip32.fromBase58(userPrv, utxolib.networks.bitcoin);
+      if (signerKeychain.isNeutered()) {
+        throw new Error('expected user private key but received public key');
+      }
+      debug(`Here is the public key of the xprv you used to sign: ${signerKeychain.neutered().toBase58()}`);
+      return signerKeychain;
+    };
 
-    if (!params.pubs || params.pubs.length !== 3) {
-      throw new Error(`must provide xpub array`);
-    }
+    let signerKeychain: utxolib.BIP32Interface | undefined;
 
-    const signerKeychain = bip32.fromBase58(userPrv, utxolib.networks.bitcoin);
-    if (signerKeychain.isNeutered()) {
-      throw new Error('expected user private key but received public key');
+    if (tx instanceof bitgo.UtxoPsbt && isTransactionWithKeyPathSpendInput) {
+      switch (params.signingStep) {
+        case 'signerNonce':
+          signerKeychain = getSignerKeychain();
+          tx.setAllInputsMusig2NonceHD(signerKeychain);
+          AbstractUtxoCoin.PSBT_CACHE.set(tx.getUnsignedTx().getId(), tx);
+          return { txHex: tx.toHex() };
+        case 'cosignerNonce':
+          assert(txPrebuild.walletId, 'walletId is required for MuSig2 bitgo nonce');
+          return { txHex: (await this.signPsbt(tx.toHex(), txPrebuild.walletId)).psbt };
+        case 'signerSignature':
+          const txId = tx.getUnsignedTx().getId();
+          const psbt = AbstractUtxoCoin.PSBT_CACHE.get(txId);
+          assert(psbt, 'psbt is missing from txCache');
+          AbstractUtxoCoin.PSBT_CACHE.delete(txId);
+          tx = psbt.combine(tx);
+          break;
+        default:
+          assert(txPrebuild.walletId, 'walletId is required for MuSig2 bitgo  nonce');
+          signerKeychain = getSignerKeychain();
+          tx.setAllInputsMusig2NonceHD(signerKeychain);
+          const response = await this.signPsbt(tx.toHex(), txPrebuild.walletId);
+          tx.combine(bitgo.createPsbtFromHex(response.psbt, this.network));
+          break;
+      }
+    } else if (params.signingStep === 'signerNonce' || params.signingStep === 'cosignerNonce') {
+      return { txHex: tx.toHex() };
```

Comment:
When does this happen? These signing steps only make sense in context of Psbt with keypath spend, no?

Reply:
In certain cases (eg: express API), the caller of this method may not know whether the txHex contains a psbt with taproot key path spend input(s).
Instead of throwing error, no-op and return the txHex. So that the caller can call this method in the same 3 sequence.